NDTensors/src/lib/Sectors/src/category_product.jl

```diff
@@ -34,78 +52,224 @@ function Base.show(io::IO, s::CategoryProduct)
   return print(io, ")")
 end
 
-category_show(io::IO, k, v) = print(io, v)
-
+category_show(io::IO, ::Int, v) = print(io, v)
 category_show(io::IO, k::Symbol, v) = print(io, "($k=$v,)")
 
+function Base.isless(s1::C, s2::C) where {C<:CategoryProduct}
+  return categories_isless(categories(s1), categories(s2))
+end
+function Base.isless(s1::CategoryProduct, s2::CategoryProduct)
+  return categories_isless(categories(s1), categories(s2))
+end
+
+# =======================================  shared  =========================================
+# there are 2 implementations for CategoryProduct
+# - ordered-like with a Tuple
+# - dictionary-like with a NamedTuple
+
+# get clean results when mixing implementations
+categories_isequal(::Tuple, ::NamedTuple) = false
+categories_isequal(::NamedTuple, ::Tuple) = false
+
+categories_isless(::NamedTuple, ::Tuple) = throw(ArgumentError("Not implemented"))
+categories_isless(::Tuple, ::NamedTuple) = throw(ArgumentError("Not implemented"))
+
+categories_type(::Type{<:CategoryProduct{T}}) where {T} = T
+
+function categories_fusion_rule(cats1, cats2)
+  diff_cat = CategoryProduct(categories_diff(cats1, cats2))
+  shared1, shared2 = categories_common(cats1, cats2)
+  fused = map(fusion_rule, values(shared1), values(shared2))
+  factorized = factorize_gradedaxes(fused)
+  type_fixed = recover_category_product_type(typeof(shared1), factorized)
+  return type_fixed × diff_cat
+end
```

Comment:
I think this refactor was helpful in simplifying this code, and does a nice job "pushing down" the complexity of the `categories_fusion_rule` function to the two functions `factorize_gradedaxes` and `recover_category_product_type` so it is easier to understand.

It appears that a lot of the complexity here is caused by wrangling various data formats, for example `Tuple` vs. `NamedTuple` storage types of `CategoryProduct`, and the fact that `fusion_rule` outputs different formats depending on whether the group is abelian or non-abelian.

One refactor I would suggest is:
```julia
function categories_fusion_rule(cats1, cats2)
  shared_cat = shared_categories_fusion_rule(categories_common(cats1, cats2)...)
  diff_cat = CategoryProduct(categories_diff(cats1, cats2))
  return shared_cat × diff_cat
end
```
so then we can focus on the implementation of `shared_categories_fusion_rule`.

I think it would simplify things to implement separate versions of `shared_categories_fusion_rule` for `Tuple` and `NamedTuple`, and write the `NamedTuple` version in terms of the `Tuple` version i.e.:
```julia
function shared_categories_fusion_rule(cats1::Tuple, cats2::Tuple)
  fused = map(fusion_rule, cats1, cats2)
  return factorize_gradedaxes(fused)
end

function shared_categories_fusion_rule(cats1::NamedTuple, cats2::NamedTuple)
  keys(cats1) == keys(cats2) || error("Keys must match")
  tuple_fused = shared_categories_fusion_rule(values(cats1), values(cats2))
  # Convert `tuple_fused` to `NamedTuple` based on `keys(cats1)`/`keys(cats2)`.
end
```

Regarding the complexity surrounding the different data formats output by `fusion_rule`, my strategy there would be to first convert everything to a common format (i.e. use `to_graded_axis` to convert the output of `fusion_rule` to `GradedOneTo`), and then at the end detect based on the inputs if the output should be unwrapped back to a single `CategoryProduct` (i.e. in those cases there should just be one block of size one, and just grab that block label), say by checking that the inputs are all `AbstractCategory` and the `SymmetryStyle` is all `AbelianGroup`.

Otherwise, perhaps it just requires different implementations based on `SymmetryStyle` for abelian and non-abelian, and making one single generic version leads to too much complexity (my suggestion above probably requires some dispatch on `SymmetryStyle` but still would try to keep more parts generic).

Hopefully that guidance isn't too vague and can at least get you started thinking about how to simplify the implementation further. @emstoudenmire I'm curious to hear your thoughts on this.

I'm very confident we can make this quite simple and elegant in the end once we navigate these conversion/data format issues. I think the current implementation of the actual fusion operation is already very nice and simple, as I said I think all of the complexity right now is in wrangling various data formats and conversions which I think can be handled in a simple and elegant way with the right generic code perspective and code organization.

Reply:
I introduced dispatch over `SymmetryStyle`. The cast to graded axis is done inside `×`.  I split `shared_categories_fusion_rule` depending on Tuple or NamedTuple.

Reply:
Please have a look at the latest version.